README
                                TilEm
                               -------

TilEm is an emulator and debugger for Texas Instruments Z80-based
graphing calculators.  It can emulate any of the following calculator
models:

    TI-73 / TI-73 Explorer
    TI-76.fr
    TI-81
    TI-82
    TI-82 STATS / TI-82 STATS.fr
    TI-83
    TI-83 Plus / TI-83 Plus Silver Edition / TI-83 Plus.fr
    TI-84 Plus / TI-84 Plus Silver Edition / TI-84 pocket.fr
    TI-85
    TI-86

TilEm fully supports all known versions of the above calculators (as
of 2012), and attempts to reproduce the behavior of the original
calculator hardware as faithfully as possible.

In addition, TilEm can emulate the TI-Nspire's virtual TI-84 Plus
mode.  This is currently experimental, and some programs may not work
correctly.

TilEm runs on the X Window System on GNU/Linux and other Unix-like
platforms, as well as on Microsoft Windows, and any other platform
supported by the GTK+ library.


                             Installation
                            --------------
Packages for Microsoft Windows will soon be available from the TilEm
project website (http://lpg.ticalc.org/prj_tilem/).  For other
platforms, you will need to compile TilEm from source; please see the
file 'INSTALL' in the source package.


                             Using TilEm
                            -------------
TilEm requires a copy of the operating system from the calculator
model(s) you wish to emulate.  This file is called a "ROM image",
since the calculator OS was traditionally stored in Read-Only Memory.
ROM images are copyrighted by TI and may not be distributed without
permission.  See the TilEm User's Manual for more information about
how to create a ROM image.

The main TilEm window shows an image of the calculator (if possible -
we are still missing background images for a few models.)  Clicking
with the left mouse button presses a key; clicking with the middle
button presses a key and holds it down.  Clicking with the right
button, or pressing Shift+F10, opens the menu.

When you run TilEm for the first time, it will ask you to select a ROM
image to use.  The file you select will be used by default the next
time you run TilEm.  You can switch to a different ROM image by
right-clicking and selecting "Open Calculator".

The state of the emulated calculator is not saved by default; to save
the state, right-click and select "Save Calculator".  The state is
saved to a ".sav" file, stored in the same directory as the ROM image.

You can send program and variable files to the emulated calculator,
either by dragging and dropping them from your file manager, or by
right-clicking and selecting "Send File".  To retrieve program or
variable files from the calculator, select "Receive File".

Other features of TilEm include:

 - A debugger for assembly programs
 - Capturing screenshots, both normal and animated
 - Recording and replaying keystroke macros

For more information, see the User's Manual.


                          About this program
                         --------------------
Many people deserve credit for helping to make this program possible.
See the file 'THANKS'.

This program is free software, which means that you are allowed to
modify it, and to distribute it (or your modified version) to others.
When you received this program, you should also have been offered a
complete copy of its source code.  For more information, see the file
'COPYING'.

You can contact the authors at <tilem-devel@lists.sourceforge.net>.
Please let us know of any problems you encounter, or ideas for how we
could make TilEm better.


NEWS
                           Version History
                          -----------------

2012-xx-xx -- version 2.1

  * French translation added.  We would be happy to support other
    languages as well - please contact us if you can help with
    translating TilEm into your native language.

  * TilEm can be connected to a real calculator (or other compatible
    device) via an external link cable, or two emulators can be
    connected together by a "virtual" cable.  Libticables2 version
    1.3.4 or later is required for external BlackLink/ParallelLink
    support.  On Windows, you must also install TiLP in order to use
    an external link cable.  Please note that there is no support yet
    for direct USB cables.

  * When you scroll up in the disassembler, it will try to find the
    start of the previous instruction.  Because Z80 instructions have
    variable length, this is heuristic and will not work correctly in
    some cases (when a section of non-executable data precedes the
    code.)  For those cases where the default scrolling behavior does
    not work correctly, you can use the "One Byte Backward" and "One
    Byte Forward" menu items to make adjustments.

  * On Windows, configuration files are stored within the TilEm
    installation directory if installed in "single user" mode (or if
    installed as a non-admin user.)

  * New skins for TI-83 Plus.fr and TI-85 (images by Olivier Jeulin.)

  * Miscellaneous bugfixes and performance improvements.


2012-06-07 -- version 2.0

  This is the first official release of the "new" TilEm.  Much of the
  old TilEm code has been rewritten, and there are many improvements
  (and probably some new bugs.)

  Please note, if you have used older versions of TilEm:

    * Your existing ROM files and settings in ~/.TilEm will not be
      used (in fact, you can keep TilEm 0.97x installed alongside
      TilEm 2.0 if you wish.)  TilEm 2.0 no longer uses a "library" of
      ROM files; you can store ROM files anywhere you like.

    * TilEm 2.0 uses a new format for calculator state (SAV) files.
      State files created by TilEm 0.97x can be loaded by TilEm 2.0,
      but if you then save the state, it will be stored in the new
      format, which older versions of TilEm will not support.

  New features and bugs fixed since version 0.975 include:

    * All code that was covered by the Z80em license has been removed.

    * Support for the TI-81 (both hardware versions) and TI-76.fr, and
      experimental support for the TI-Nspire's TI-84 Plus emulation
      mode.

    * Many hardware emulation improvements for all calculator models.
      In particular, major improvements have been made concerning Z80
      interrupts, timers, the LCD driver, and the link port.

    * The emulator window uses TiEmu-format skin files.

    * Greatly improved grayscale emulation.

    * Commands for saving still screenshots (in PNG, BMP, JPEG, or GIF
      format) and animations (GIF format only.)

    * Keypad macros can be recorded and replayed.

    * Programs and/or ROM files can be loaded from the command line.

    * Link I/O uses libticalcs2, which allows all types of variables,
      as well as Flash apps and OSes, to be transferred through the
      link port.  For the TI-81, PRG files can be transferred to and
      from the calculator memory directly.

    * TilEm does not consume 100% of the host CPU when idle.

    * Improved disassembler (macros; distinct "labels" and "romcalls";
      named IY flags.)

    * The debugger offers a "Finish Subroutine" command.  In addition,
      the "Step Over" command behaves more sensibly.

    * Breakpoints can be set on absolute memory addresses, and on Z80
      opcodes.

    * Many minor improvements.

  Features of 0.975 that are not yet supported in TilEm 2.0 include:

    * External link cables.

    * Custom symbol files in the disassembler.

    * Program counter history tracking.

  Most of the new code is due to Benjamin Moody (floppusmaximus) and
  Thibault Duponchelle (contra-sh).  See THANKS for a full list of
  contributors.


core/rom.c
/*
 * libtilemcore - Graphing calculator emulation library
 *
 * Copyright (C) 2009 Benjamin Moody
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdio.h>
#include <string.h>
#include "tilem.h"

static int find_string(const char *str, FILE *romfile,
		       dword start, dword limit)
{
	char buf[256];
	int pos = 0;
	int len, i;

	len = strlen(str);

	fseek(romfile, (long int) start, SEEK_SET);

	for (i = 0; i < len-1; i++) {
		buf[pos] = fgetc(romfile);
		pos = (pos+1)%256;
		limit--;
	}

	while (limit > 0 && !feof(romfile) && !ferror(romfile)) {
		buf[pos] = fgetc(romfile);
		pos = (pos+1)%256;
		limit--;

		for (i = 0; i < len; i++) {
			if (str[i] != buf[(pos + 256 - len + i)%256])
				break;
		}
		if (i == len)
			return 1;
	}
	return 0;
}

char tilem_guess_rom_type(FILE* romfile)
{
	unsigned long initpos;
	dword size;
	char result;

	initpos = ftell(romfile);

	fseek(romfile, 0L, SEEK_END);
	size = ftell(romfile);

	if (size >= 0x8000 && size < 0x9000) {
		/* 32k: TI-81 (old or new) */
		result = TILEM_CALC_TI81;
	}
	else if (size >= 0x20000 && size < 0x2C000) {
		/* 128k: TI-82 or TI-86 */
		if (find_string("CATALOG", romfile, 0, 0x20000))
			result = TILEM_CALC_TI85;
		else
			result = TILEM_CALC_TI82;
	}
	else if (size >= 0x40000 && size < 0x4C000) {
		/* 256k: TI-83 (or a variant) or TI-86 */
		if (!find_string("TI82", romfile, 0, 0x40000))
			result = TILEM_CALC_TI86;
		else if (find_string("Termin\x96", romfile, 0, 0x40000))
			result = TILEM_CALC_TI76;
		else
			result = TILEM_CALC_TI83;
	}
	else if (size >= 0x80000 && size < 0x8C000) {
		/* 512k: TI-83 Plus or TI-73 */
		if (find_string("TI-83 Plus", romfile, 0, 8 * 0x4000))
			result = TILEM_CALC_TI83P;
		else
			result = TILEM_CALC_TI73;
	}
	else if (size >= 0x100000 && size < 0x124000) {
		/* 1024k: TI-84 Plus */
		result = TILEM_CALC_TI84P;
	}
	else if (size >= 0x200000 && size < 0x224000) {
		/* 2048k: TI-83 Plus SE, TI-84 Plus SE */
		if (find_string("Operating", romfile, 0x1FC000, 0x4000))
			result = TILEM_CALC_TI84P_SE;
		else
			result = TILEM_CALC_TI83P_SE;
	}
	else {
		result = 0;
	}

	fseek(romfile, initpos, SEEK_SET);
	return result;
}
